File: sicp_1_3.swift
#!/usr/bin/env swift
// sicp exercise 1.3

// Exercise 1.3.  Define a procedure that takes three numbers as arguments and returns the sum of 
// the squares of the two larger numbers. 

func sumOfSquaresOfTwoLargerInts(numbers: [Int]) -> Int {
  println(numbers)
  if numbers.count == 2 {
    return numbers.map { square($0) }.reduce(0) { $0 + $1 }
  } else {
    if maxElement(numbers) == minElement(numbers) {
      return sumOfSquaresOfTwoLargerInts([numbers[0],numbers[1]])
    } else {
      let newNumbers = numbers.filter { $0 > minElement(numbers) }
      return sumOfSquaresOfTwoLargerInts(newNumbers)
    }
  }
}

func square(n: Int) -> Int {
  return n * n
}

println(sumOfSquaresOfTwoLargerInts([1,2,3]))
println(sumOfSquaresOfTwoLargerInts([3,3,3]))

File: sicp_1_3.js
#!/usr/bin/env node
// sicp exercise 1.3 in node

function square(x) { return x * x }
function sum(x,y) { return x + y }
function sumSquares(x,y) { return sum(square(x),square(y)) }
function maximum(x,y) { return x > y ? x : y }
function sumTwoLargerSquares(a,b,c) { return sumSquares(maximum(a,b),maximum(b,c)) }
console.log(sumTwoLargerSquares(1,2,3))
console.log(sumTwoLargerSquares(2,1,1))
